analyzer/topology_agent_endpoint.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2016 Red Hat, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package analyzer
+
+import (
+	"sync"
+
+	shttp "github.com/skydive-project/skydive/http"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology/graph"
+)
+
+// TopologyAgentEndpoint serves the graph for agents.
+type TopologyAgentEndpoint struct {
+	sync.RWMutex
+	shttp.DefaultWSSpeakerEventHandler
+	pool   shttp.WSJSONSpeakerPool
+	Graph  *graph.Graph
+	cached *graph.CachedBackend
+	wg     sync.WaitGroup
+}
+
+// OnDisconnected called when an agent disconnected.
+func (t *TopologyAgentEndpoint) OnDisconnected(c shttp.WSSpeaker) {
+	host := c.GetHost()
+	t.Graph.Lock()
+	logging.GetLogger().Debugf("Authoritative client unregistered, delete resources %s", host)
+	t.Graph.DelHostGraph(host)
+	t.Graph.Unlock()
+}
+
+func (t *TopologyAgentEndpoint) hostGraphDeleted(host string, mode int) {
+	t.cached.SetMode(mode)
+	defer t.cached.SetMode(graph.DefaultMode)
+
+	t.Graph.DelHostGraph(host)
+}
+
+// OnWSJSONMessage is triggered when a message from the agent is received.
+func (t *TopologyAgentEndpoint) OnWSJSONMessage(c shttp.WSSpeaker, msg *shttp.WSJSONMessage) {
+	msgType, obj, err := graph.UnmarshalWSMessage(msg)
+	if err != nil {
+		logging.GetLogger().Errorf("Graph: Unable to parse the event %v: %s", msg, err.Error())
+		return
+	}
+
+	t.Graph.Lock()
+	defer t.Graph.Unlock()
+
+	switch msgType {
+	case graph.HostGraphDeletedMsgType:
+		// HostGraphDeletedMsgType is handled specifically as we need to be sure to not use the
+		// cache while deleting otherwise the delete mechanism is using the cache to walk throught
+		// the graph.
+		logging.GetLogger().Debugf("Got %s message for host %s", graph.HostGraphDeletedMsgType, obj.(string))
+		t.hostGraphDeleted(obj.(string), graph.CacheOnlyMode)
```

Comment:
simply t.Graph.DelHostGraph(host) here instead of two calls

Reply:
done